### DISCUSSION TITLE
Why is error_reporting stricter for lazy services ?

### DISCUSSION THREAD

**Discussion by vctls** [symfony/symfony#47572] [Q&A] [answered]
The kernel silences include errors during some of the container initilalization ([here](https://github.com/symfony/symfony/blob/1159d133130c1474f16c3a5f44b6c54a0a81e6d6/src/Symfony/Component/HttpKernel/Kernel.php#L391)), but then re-enables it before lazy services are loaded. Why is that ?

I noticed because of an edge case with legacy code where an invalid import in a service wasn't detected until the service was declared `lazy`.

**Comment by stof** [comment, marked as answer]
The goal is only to silence include warnings, because there is no way to avoid warnings in case of race conditions between multiple requests warming up the cache.

**Reply by vctls** [reply]
Thank you, @stof! So this does not happen with lazy services?
What would be the recommended practice to enforce correct includes in non-lazy services? Or at least trace invalid ones?
btw if this is explained somewhere in the documentation, I'll take a simple link as an answer. I don't want to bother you too much.

**Reply by stof** [reply]
What do you mean "includes in non-lazy services" ?

**Reply by vctls** [reply]
I have a class with some legacy code where there is a dynamic include. This class is declared as a service. When the included file cannot be found during initialization, no errors are thrown, since these errors are explicitely silenced.  
Unless the service is declared as lazy, at which point the `error_reporting` is back to `-1`. That's when I noticed the include was failing, since they weren't even explicitely logged.  

Why are the errors not silenced when loading lazy services?  
And how should I handle failing includes when the errors are silenced? Just take the return value of include and implement my own error handler, maybe?

**Reply by ahundiak** [reply]
It might be useful if you could create a minimal skeleton project with a simple example service showing the problem.  Very simple.  Add a command to trigger the error.  Check your project in to a repo and provide a link.  Sometimes it's easier to just look at a bit of code than it is to describe it.  Plus sometimes when you try to isolate a problem it turns out that something else is going on.

**Reply by vctls** [reply]
OK, this is actually much more confusing than I thought. I'm trying to follow what happens during container initialization using xdebug, but switching just that one service to lazy actually seems to have consequences on the loading process of other classes.
I tried to reproduce the issue on a new project but it didn't work.
The problem is that the service at issue is within a vendor, among a slew of other services, and I'm having a hard time keeping track of the interactions between each of them.

**Comment by vctls** [comment]
Marking this as answered because I don't have the time right now to make a reproducible example.
